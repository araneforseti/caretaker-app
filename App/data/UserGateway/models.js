const Permissions = {
    FOCUS: 'focus',
    ADMIN: 'admin',
    HELPER: 'helper'
}

const NotificationTypes = {
    PUSH: 'push',
    EMAIL: 'email',
    SMS: 'sms'
}

const NotificationTriggers = {
    ANNOUNCEMENT: 'announcement',
    EVENT: 'event',
    EVENT_BY_ROLE: 'event by role'
}

const Settings = {
    [NotificationTypes.PUSH]: {
        name: 'Push',
        value: true
    },
    [NotificationTypes.EMAIL]: {
        name: 'Email',
        value: false
    },
    [NotificationTypes.SMS]: {
        name: 'SMS',
        value: false
    },
    [NotificationTriggers.ANNOUNCEMENT]: {
        name: 'Announcements',
        value: true
    },
    [NotificationTriggers.EVENT]: {
        name: 'Events',
        value: false
    },
    [NotificationTriggers.EVENT_BY_ROLE]: {
        name: 'Events specific to your role',
        value: true
    }
}

module.exports = {
  Permissions: Permissions,
  NotificationTypes: NotificationTypes,
  NotificationTriggers: NotificationTriggers,
  Settings: Settings
}
