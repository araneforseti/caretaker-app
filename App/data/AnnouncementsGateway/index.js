/*
  Makes API calls to fetch announcements.

  However, until there is an API to call, it returns canned data.
*/

import {withSubscribers} from '../../utils/subscribers.js';

const createAnnouncementsGateway = (subscribers) => {
    let allAnnouncements = null;

    return {
        getAll: () => {
            if(allAnnouncements === null) {
                //api call to get roles
                allAnnouncements = [
                    {id: 1, date: '2017-04-30T19:00:00-05:00', title: '1st Treatment Success', description: 'Woo, go us'},
                    {id: 2, date: '2017-05-04T19:00:00-05:00', title: '2nd Treatment Delayed', description: 'Boo'},
                    {id: 3, date: '2017-05-06T19:00:00-05:00', title: 'Need Help', description: 'Send lemons'},
                    {id: 4, date: '2017-05-06T19:00:00-05:00', title: 'I have a need', description: 'It\'s very complicated, and I\'m not sure exactly how to put it into words, but if I had to approximate it, I would probably first request a piping cup of tea and some crackers, as it would take some time and I would be loath to do my description a disservice by unduly interrupting it with breaks for necessaries.\n\nThen, with the implements of storytelling in hand, I would proceed to warm up with an oh-so-brief description of the setting, of that hot summer day on which my conundrum began. It\'s very important for you, or anyone else who listens to my tale, to understand the environment in which these events took place, otherwise true understanding will evade you. I would say that in order to transfer the feelings and emotions that I have built up around these happenings, the listener themselves must prepare to embark on a journey, and thus procure for themselves their own tea.'}
                ]
            }
            return allAnnouncements;
        },

        create: (data) => {
            //api call to create announcement
            const newId = allAnnouncements[allAnnouncements.length - 1].id + 1;
            const result = {id: newId, date: '2017-05-07T19:00:00-05:00', title: 'New Announcement ' + newId, description: 'Send lemons'}
            allAnnouncements.push(result);
            subscribers.notify();
        }
    }
}

export default withSubscribers(createAnnouncementsGateway);
