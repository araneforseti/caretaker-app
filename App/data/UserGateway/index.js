import { Record } from 'immutable';
import UserModels from './models.js'
import { API_URL } from '../../config/ApiConfig';

export const Permissions = UserModels.Permissions
export const NotificationTypes = UserModels.NotificationTypes
export const NotificationTriggers = UserModels.NotificationTriggers

const createUserSettings = () => ({
  [NotificationTypes.PUSH]: {
    name: 'Push',
    value: true
  },
  [NotificationTypes.EMAIL]: {
    name: 'Email',
    value: false
  },
  [NotificationTypes.SMS]: {
    name: 'SMS',
    value: false
  },
  [NotificationTriggers.ANNOUNCEMENT]: {
    name: 'Announcements',
    value: true
  },
  [NotificationTriggers.EVENT]: {
    name: 'Events',
    value: false
  },
  [NotificationTriggers.EVENT_BY_ROLE]: {
    name: 'Events specific to your role',
    value: true
  }
})

export const User = Record({
  name: 'Defaul User',
  identifier: 'default@domain.com',
  permission: Permissions.HELPER,
  roles: [],
  phone: '123-456-7890',
  settings: createUserSettings()
});

const createUserGateway = function() {
  let allUsers = null;
  let currentUser = null;

  return {
    getCurrentUser() {
      return currentUser;
    },

    getAll() {
      if (allUsers === null) {
        //api call goes here
        allUsers = [
          new User({ name: 'S', identifier: 's', permission: Permissions.FOCUS, roles: [], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Sarah', identifier: 'sarah@emailprovider.com', permission: Permissions.FOCUS, roles: [], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Caroline', identifier: 'caroline@woahdude.com', permission: Permissions.ADMIN, roles: ['driver', 'coordinator'], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Jack', identifier: 'jack@coolwebsite.com', permission: Permissions.HELPER, roles: ['active friend'], settings: createUserSettings() }),
          new User({ name: 'Austin', identifier: 'austin@yeehaw.com', permission: Permissions.HELPER, roles: [], phone: '123-456-7890', settings: createUserSettings() })
        ]
      }
      return allUsers;
    },

    getName(identifier) {
      const user = allUsers.find((user) => user.identifier === identifier)
      return user ? user.name : 'TBD';
    },

    getRoles() {
      return currentUser.roles
    },

    login(email) {
      //make call to google oauth and store the results
      let results = {
        email: email
      };
      let userFromBackend = this.getAll().find((user) => user.identifier === results.email);
      if (typeof userFromBackend !== 'undefined') {
        currentUser = userFromBackend;
        return true;
      }
      return false;
    },

    logout() {
      // make call to oauth and/or invalidate current session
      this.currentUser = null;
    },

    isSelf(identifier) {
      return currentUser.identifier === identifier;
    },

    isAdmin() {
      return currentUser.permission === Permissions.ADMIN ||
        currentUser.permission === Permissions.FOCUS;
    },

    getCurrentPermissions() {
      return currentUser.permission;
    },

    inviteUser(email) {
      //api call to invite the user
      alert('invited ' + email + '!');
    },

    updateSettings(settings) {
      //api call to update the user settings
    }
  }
}

export default createUserGateway();
