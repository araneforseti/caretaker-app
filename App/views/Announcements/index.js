import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  FlatList,
  TouchableOpacity
} from 'react-native';
import ContainerFor from './container.js';
import { renderIf } from '../../utils/helpers';
import ActionButton from '../../components/material-ui/ActionButton';
import ScreenWithToolbar from '../../components/shared/ScreenWithToolbar';

const ListItem = (props) => (
    <TouchableOpacity style={styles.announcementContainer} onPress={props.onEditAnnouncement}>
        <Text style={styles.announcementTitle}>{props.title}</Text>
        <Text>{new Date(props.date).toDateString()}</Text>
        <Text>{props.description}</Text>
    </TouchableOpacity>
)

const Announcements = (props) => {

    const _renderItem = ({item}) => (
        <ListItem
            title={item.title}
            date={item.date}
            description={item.description}
            onEditAnnouncement={props.onEditAnnouncement}
        />
    )

    const addAnnouncementButton = renderIf(props.isUserAdmin, () =>
        <ActionButton onPress={props.onAddAnnouncement} />
    );

    return <ScreenWithToolbar title='Announcements' navigation={props.navigation}>
        <View style={styles.container}>
            <Text style={styles.title}>Announcements</Text>
            <FlatList
               data={props.announcements}
               renderItem={_renderItem}
               keyExtractor={(item) => item.id}
               removeClippedSubviews={true}
            />
            {addAnnouncementButton}
        </View>
    </ScreenWithToolbar>;
}

export default ContainerFor(Announcements);

const styles = StyleSheet.create({
  button: {
    margin: 10,
    padding: 10
  },
  container: {
    flex: 1,
    backgroundColor: '#F5FCFF',
  },
  title: {
    fontSize: 24,
    textAlign: 'center',
    margin: 15,
  },
  announcementContainer: {
      margin: 10
  },
  announcementTitle: {
      fontSize: 16,
      fontWeight: 'bold',
      marginTop: 10,
  }
});
