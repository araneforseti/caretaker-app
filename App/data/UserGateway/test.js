import UserGateway from './index.js';
import { Record } from 'immutable';

describe('UserGateway', () => {

    it('should return false when attempting to sign in with a non-existent user', () => {
        expect(UserGateway.login('TBD')).toBe(false);
    });

    it('should return true when sign in is successful', () => {
        expect(UserGateway.login('sarah@emailprovider.com')).toBe(true);
    });

    it('should assign TBD as a name when TBD is the identifier', () => {
        expect(UserGateway.getName('TBD')).toBe('TBD');
    });

    it('should get the current user as an immutable record', () => {
        UserGateway.login('s');

        const user =  UserGateway.getCurrentUser();

        //Record.isRecord is in the immutable@4.0.0 release candidate
        expect(user instanceof Record).toBe(true);
    })
});
