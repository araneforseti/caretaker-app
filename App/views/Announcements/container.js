import React, { Component } from 'react';
import AnnouncementsGateway from '../../data/AnnouncementsGateway/';
import UserGateway from '../../data/UserGateway/';

export default ContainerFor = (Announcements) => class extends Component {
    constructor() {
        super();
        this.state = {
            announcements: [],
            isUserAdmin: false
        }
    }

    componentDidMount() {
        this.setState({
            announcements: AnnouncementsGateway.getAll(),
            isUserAdmin: UserGateway.isAdmin()
        });
        AnnouncementsGateway.subscribe(this.updateAnnouncements);
    }

    componentWillUnmount() {
        AnnouncementsGateway.unsubscribe(this.updateAnnouncements);
    }

    updateAnnouncements = () => {
        const announcements = AnnouncementsGateway.getAll()
        let newState = this.state;
        newState.announcements = announcements;
        this.setState(newState);
    }

    _onAddAnnouncement = () => {
        AnnouncementsGateway.create({some: 'data'})
    }

    _onEditAnnouncement = () => {
        if(this.state.isUserAdmin) {
            alert('announcement edited!')
        }
    }

    render() {
        return (
            <Announcements
                announcements={this.state.announcements}
                isUserAdmin={this.state.isUserAdmin}
                onAddAnnouncement={this._onAddAnnouncement}
                onEditAnnouncement={this._onEditAnnouncement}
            />
        )
    }
}
