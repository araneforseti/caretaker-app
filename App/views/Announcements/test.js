import React, {Component} from 'react';
import renderer from 'react-test-renderer';
import AnnouncementsGateway from '../../data/AnnouncementsGateway/';
import UserGateway from '../../data/UserGateway/';
import Announcements from './index.js';

describe('Announcements Screen', () => {

    beforeEach(() => {
        jest.resetAllMocks();
        UserGateway.signIn('sarah@emailprovider.com');

        Date.prototype.toDateString = jest.fn(() => 'fake date');
    });

    it('snapshot', () => {
        const tree = renderer.create(<Announcements />);

        expect(tree).toMatchSnapshot();
    })

    it('should get all announcements from the gateway', () => {
        const spy = jest.spyOn(AnnouncementsGateway, 'getAll');

        renderer.create(<Announcements />);

        expect(spy).toHaveBeenCalled();
    });

    it('should subscribe to the announcements gateway', () => {
        const spy = jest.spyOn(AnnouncementsGateway, 'subscribe');

        const announcements = renderer.create(<Announcements />).getInstance();

        expect(spy).toHaveBeenCalledWith(announcements.updateAnnouncements);
    });

    it('should update all the announcements', () => {
        const oneAnnouncement = [{id: 1, title: 'Announcement'}];
        const twoAnnouncements = [{id: 1, title: 'Announcement'}, {id: 2, title: 'Announcement'}];
        jest.spyOn(AnnouncementsGateway, 'getAll')
            .mockReturnValueOnce(oneAnnouncement)
            .mockReturnValueOnce(twoAnnouncements)
        const announcements = renderer.create(<Announcements />).getInstance();

        announcements.updateAnnouncements();

        expect(announcements.state.announcements).toEqual(twoAnnouncements);
    })

    it('should show an action button when the user is an admin or focus', () => {
        jest.spyOn(UserGateway, 'isAdmin')
            .mockReturnValue(true);

        const announcements = renderer.create(<Announcements />).getInstance();

        expect(announcements.state.isUserAdmin).toBe(true);
    });

    it('should add an announcement when the add announcement button is pressed', () => {
        const spy = jest.spyOn(AnnouncementsGateway, 'create');
        jest.spyOn(UserGateway, 'isAdmin')
            .mockReturnValue(true);
        const announcements = renderer.create(<Announcements />).getInstance();

        announcements._onAddAnnouncement();

        expect(spy).toHaveBeenCalledWith({some: 'data'});
    });

    it('should edit an announcement when the user is an admin', () => {
        const spy = jest.spyOn(window, 'alert');
        jest.spyOn(UserGateway, 'isAdmin')
            .mockReturnValue(true);
        const announcements = renderer.create(<Announcements />).getInstance();

        announcements._onEditAnnouncement();

        expect(spy).toHaveBeenCalledWith('announcement edited!');
    });

    it('should not edit an announcement when the user is not an admin', () => {
        const spy = jest.spyOn(window, 'alert');
        jest.spyOn(UserGateway, 'isAdmin')
            .mockReturnValue(false);
        const announcements = renderer.create(<Announcements />).getInstance();

        announcements._onEditAnnouncement();

        expect(spy).not.toHaveBeenCalled();
    })
})
