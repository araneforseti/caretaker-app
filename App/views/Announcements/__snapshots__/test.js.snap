// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Announcements Screen snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 24,
        "margin": 15,
        "textAlign": "center",
      }
    }
  >
    Announcements
  </Text>
  <RCTScrollView
    data={
      Array [
        Object {
          "date": "2017-04-30T19:00:00-05:00",
          "description": "Woo, go us",
          "id": 1,
          "title": "1st Treatment Success",
        },
        Object {
          "date": "2017-05-04T19:00:00-05:00",
          "description": "Boo",
          "id": 2,
          "title": "2nd Treatment Delayed",
        },
        Object {
          "date": "2017-05-06T19:00:00-05:00",
          "description": "Send lemons",
          "id": 3,
          "title": "Need Help",
        },
        Object {
          "date": "2017-05-06T19:00:00-05:00",
          "description": "It's very complicated, and I'm not sure exactly how to put it into words, but if I had to approximate it, I would probably first request a piping cup of tea and some crackers, as it would take some time and I would be loath to do my description a disservice by unduly interrupting it with breaks for necessaries.
      
      Then, with the implements of storytelling in hand, I would proceed to warm up with an oh-so-brief description of the setting, of that hot summer day on which my conundrum began. It's very important for you, or anyone else who listens to my tale, to understand the environment in which these events took place, otherwise true understanding will evade you. I would say that in order to transfer the feelings and emotions that I have built up around these happenings, the listener themselves must prepare to embark on a journey, and thus procure for themselves their own tea.",
          "id": 4,
          "title": "I have a need",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onViewableItemsChanged={undefined}
    removeClippedSubviews={true}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "margin": 10,
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "bold",
                "marginTop": 10,
              }
            }
          >
            1st Treatment Success
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            fake date
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            Woo, go us
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "margin": 10,
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "bold",
                "marginTop": 10,
              }
            }
          >
            2nd Treatment Delayed
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            fake date
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            Boo
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "margin": 10,
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "bold",
                "marginTop": 10,
              }
            }
          >
            Need Help
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            fake date
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            Send lemons
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "margin": 10,
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "bold",
                "marginTop": 10,
              }
            }
          >
            I have a need
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            fake date
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            It's very complicated, and I'm not sure exactly how to put it into words, but if I had to approximate it, I would probably first request a piping cup of tea and some crackers, as it would take some time and I would be loath to do my description a disservice by unduly interrupting it with breaks for necessaries.
            
            Then, with the implements of storytelling in hand, I would proceed to warm up with an oh-so-brief description of the setting, of that hot summer day on which my conundrum began. It's very important for you, or anyone else who listens to my tale, to understand the environment in which these events took place, otherwise true understanding will evade you. I would say that in order to transfer the feelings and emotions that I have built up around these happenings, the listener themselves must prepare to embark on a journey, and thus procure for themselves their own tea.
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
  <ActionButton
    onPress={[Function]}
  />
</View>
`;
